### PR TITLE
Implement SIA-R87

### PR DESCRIPTION
**Pull request by kasperisager**

### CODE REVIEW

**Review by Jym77 — CHANGES_REQUESTED** [review]
Awesome work!

File: packages/alfa-rules/src/common/predicate/is-keyboard-actionable.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import { Context } from "@siteimprove/alfa-selector";
+
+import { isVisible } from "./is-visible";
+import { isTabbable } from "./is-tabbable";
+
+const { and } = Predicate;
+
+export function isKeyboardActionable(device: Device): Predicate<Element> {
+  return and(isTabbable(device), (element) =>
+    isVisible(device, Context.focus(element))(element)
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/common/predicate/is-keyboard-actionable.ts:13]: We need to pass `element` to `Context.focus()` so we can't shorten this further I'm afraid.

[the remark this answers is not in the document]
Comment (Jym77) [reply packages/alfa-rules/src/common/predicate/is-keyboard-actionable.ts:13]: True 🤦

File: packages/alfa-rules/src/sia-r87/rule.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  uri: "https://siteimprove.github.io/sanshikan/rules/sia-r87.html",
+  evaluate({ device, document, response }) {
+    return {
+      applicability() {
+        return fold(
+          hasChild(isDocumentElement),
+          () => [document],
+          () => [],
+          document
+        );
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/sia-r87/rule.ts:32]: I much prefer avoiding the `fn(a)(b)` invocation, which isn't common and makes things harder grok. When the `fn(a)` bit becomes sufficiently long, it's easy to miss that the resulting function is invoked by `(b)`. It's the reason why `Predicate.test()` and `Predicate.fold()` exist; to clearly separate the predicate from its operand.

File: packages/alfa-rules/src/sia-r87/rule.ts [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            () => Outcomes.HasTabbable,
+            () => Outcomes.HasNoTabbable
+          ),
+
+          2: expectation(
+            element.isNone(),
+            () => Outcomes.HasNoTabbable,
+            () =>
+              expectation(
+                element.none(hasRole((role) => role.is("link"))),
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/sia-r87/rule.ts:74]: That's already baked into `Role#is()` 👍

File: packages/alfa-cascade/src/user-agent.ts [hunk 4]
```diff
@@ -51,7 +51,7 @@ export const UserAgent = h.sheet([
   }),
 
   /**
-   * @see https://html.spec.whatwg.org/#non-replaced-elements-flow-content
+   * @see https://html.spec.whatwg.org/#flow-content-3
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-cascade/src/user-agent.ts:54]: It's not a direct translation of the suggested rendering hints and so auto-generating it wouldn't really be appropriate.

File: packages/alfa-rules/src/common/predicate/is-keyboard-actionable.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import { Element } from "@siteimprove/alfa-dom";
+import { Predicate } from "@siteimprove/alfa-predicate";
+import { Context } from "@siteimprove/alfa-selector";
+
+import { isVisible } from "./is-visible";
+import { isTabbable } from "./is-tabbable";
+
+const { and } = Predicate;
+
+export function isKeyboardActionable(device: Device): Predicate<Element> {
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/common/predicate/is-keyboard-actionable.ts:11]: It's indeed an implementation assumption that things that can be tabbed to can also be activated by keyboard. The issue is really that we can never surely say if an element can be activated by keyboard even when it appears to be the case, so we either _always_ pose a question or we rely entirely on assumptions.

I'll add a remark.

File: packages/alfa-rules/src/sia-r87/rule.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+          () => [],
+          document
+        );
+      },
+
+      expectations(target) {
+        const element = target.tabOrder().find(isTabbable(device));
+
+        const url = element
+          .filter(hasName("a"))
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/sia-r87/rule.ts:39]: `<area>` would be the only other one we'd be able to consider, though I'm not sure why anyone would create a skip link using an `<area>` element. It doesn't really cost us anything to support that though so why not 😄

File: packages/alfa-rules/src/sia-r87/rule.ts [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+                    element.none(not(isIgnored(device))),
+                    () => Outcomes.FirstTabbableIsIgnored,
+                    () =>
+                      expectation(
+                        element.none(isKeyboardActionable(device)),
+                        () => Outcomes.FirstTabbableIsNotKeyboardActionable,
+                        () =>
+                          reference.isSome()
+                            ? expectation(
+                                reference.some(hasRole("main")),
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/sia-r87/rule.ts:87]: If the element pointed to doesn't have a role of `main`, the rule will ask if the element referenced is the main content. We could easily automate the case where the reference is at the start of an element with a role of `main` though; I'll leave that as a to-do.

File: packages/alfa-rules/src/sia-r87/rule.ts [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+                () =>
+                  expectation(
+                    element.none(not(isIgnored(device))),
+                    () => Outcomes.FirstTabbableIsIgnored,
+                    () =>
+                      expectation(
+                        element.none(isKeyboardActionable(device)),
+                        () => Outcomes.FirstTabbableIsNotKeyboardActionable,
+                        () =>
+                          reference.isSome()
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/sia-r87/rule.ts:85]: Questions are values, and immutable ones at that, so we can just pull them into the outer scope and reuse them. The issue is that when we don't know where the link points, we need confirmation first that the link is indeed internal.

**Review by Jym77 — APPROVED** [review]

File: packages/alfa-rules/src/sia-r87/rule.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jym77) [reply packages/alfa-rules/src/sia-r87/rule.ts]: I guess there could be an image map with links to sections of content 🤷

File: packages/alfa-rules/src/sia-r87/rule.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Jym77) [reply packages/alfa-rules/src/sia-r87/rule.ts:108]: Yeah, I  feel we need more data before really acting on that…
Having a auto "at the start" detection will save questions in some cases (where the link goes "at the start" but not on an element with a role of main). But we can probably wait and see how frequent these cases are before deciding whether to act.

**Pull request merged**